### PR TITLE
Fix UrlInput combobox to use the ARIA 1.0 pattern.

### LINKED ISSUE #47147: UrlInput: Stick to the ARIA 1.0 pattern
### Description

I was wondering why Safari + VoiceOver fail reading the options of the UrlInput combobox while they do announce the options of the Tag suggestions. Screenshot:

Screenshot 2023-01-13 at 16 14 47

Turns out the input field of UrlInput uses `aria-controls` while the one of the Tag suggestions uses `aria-owns`. This was changed in https://github.com/WordPress/gutenberg/pull/43278 and only for the UrlInput component.

Please notice that making a change that may look trivial can actually have a huge impact on assistive technology. Also, Safari + VoiceOver are known to be buggy with comboboxes.

It's very, very, important to _actually test_ the impact of such changes on the most common browsers / screen reader combinations. Even the most simple changes need to be testsd.

## TL;DR

The change in https://github.com/WordPress/gutenberg/pull/43278 was made following on [this comment](https://github.com/WordPress/gutenberg/pull/43147#discussion_r945498693), based on a recommendation on the [MDN combobox documentation](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/combobox_role):

> Note: In older ARIA specs, this was aria-owns rather than aria-controls, so you may see aria-owns in older combobox implementations. The aria-owns in the code should be updated to aria-controls!

Unfortunately, that's inaccurate. 

It is true the newer ARIA specs recommends `aria-controls`. However, the combobox ARIA pattern evolved over time across the ARIA 1.0, 1.1, and 1.2 specs. It is important to note that the HTML structure itself changed, together with the usage of some aria-properties. To recap:

### ARIA 1.2
It's still a 'Candidate Recommendation Draft', stuck on 08 December 2021. As such, it's not a official recommendation and it's not guaranteed to be implemented by vendors. 
It provides an interesting note at https://www.w3.org/TR/wai-aria-1.2/#combobox

> The structural requirements for combobox defined by this version of the specification are different from the requirements defined by ARIA 1.0 and ARIA 1.1:
> - The ARIA 1.0 specification required the input element with the combobox role to be a single-line text field and reference the popup element with aria-owns instead of aria-controls.
> - The ARIA 1.1 specification, which was not broadly supported by assistive technologies, required the combobox to be a non-focusable element with two required owned elements -- a focusable textbox and a popup element controlled by the textbox.
> - The changes introduced in ARIA 1.2 improve interoperability with assistive technologies and enable authors to create presentations of combobox that more closely imitate a native HTML select element.

Here's the HTML example in ARIA 1.2:

```
<input type="text" id="tag_combo"
      role="combobox" aria-autocomplete="list"
      aria-haspopup="listbox" aria-expanded="true"
      aria-controls="popup_listbox" aria-activedescendant="selected_option">
<ul role="listbox" id="popup_listbox">
   <li role="option"> ...
```

Again, we can't use this pattern.

### ARIA 1.1
ARIA 1.1 is the current official recommendation, dated 14 December 2017.
As mentioned in ARIA 1.2, the 1.1 combobox pattern was not broadly supported by assistive technologies, Basically it doesn't work well and we can't use it.

Here's [the HTML example in ARIA 1.1](https://www.w3.org/TR/wai-aria-1.1/#combobox), where a div element 'owns' both the input and the listbox (via nesting and aria-owns):

```
<div aria-label="Tag" role="combobox" aria-expanded="true" aria-owns="owned_listbox" aria-haspopup="listbox">
    <input type="text" aria-autocomplete="list" aria-controls="owned_listbox" aria-activedescendant="selected_option">
</div>
<ul role="listbox" id="owned_listbox">
    <li role="option">Zebra</li>
    <li role="option" id="selected_option">Zoom</li>
</ul>
```

Again, we can't use this pattern.

### ARIA 1.0 
This is an old pattern but it's the only one that is currently broadly supported. It uses `aria-owns` in the input field.

Here's [the HTML example in ARIA 1.0](https://www.w3.org/TR/wai-aria-1.0/complete#combobox):

```
<input type="text" aria-label="Tag" role="combobox" aria-expanded="true"
  aria-autocomplete="list" aria-owns="owned_listbox" aria-activedescendant="selected_option">
<ul role="listbox" id="owned_listbox">
  <li role="option">Zebra</li>
  <li role="option" id="selected_option">Zoom</li>
</ul>
```

By restoring `aria-owns`, Safari + VoiceOver magically announce the options again.

## Aside
There's one more detail that needs to be improved.
The `aria-selected` attribute needs to be rendered only on the 'highlighted' option and omitted from all the other options. The only allowed usage of `aria-selected` with a value of `false` in an ARIA listbox is when the Listbox allows selcting multiple items.

To clarify, here's a screenshot from the W3C ARIA Authoring Practices [Editable Combobox with List Autocomplete](https://www.w3.org/WAI/ARIA/apg/example-index/combobox/combobox-autocomplete-list.html):

Screenshot 2023-01-13 at 16 47 21

### Step-by-step reproduction instructions

- edit a post
- use Safari and VoiceOver
- add a link to a word in a paragraph
- type 'hello' in the URL input search to get some link suggestions
- use the Down and Up keys do navigate through the suggestion options
- observe VoiceOver doesn't announce the options
- make sure the settings sidebar displays the Post tab
- click to add a Tag to the post
- type something in the 'Add new tag' field to trigger the list of suggestions
- observe VoiceOver does announce the options

Note: as mentioned above, Safari and VoiceOver are buggy anyways. When testing, you may notice some weirdness also when using the Tag selector. They announce things differently or in a buggy way e.g.:
- sometimes they just announce `{option name}, clickable`
- sometimes they announce `{option name}, selected, (1 of n), completion selected` (this is the expected behavior)
- sometimes you may need to delete one of the typed characters and re-type it again to make VoiceOver announce the options

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by afercia**
## What?

Fixes #47147
Safari and VoiceOver don't read the UrlInput combobox suggestions. To make the combobox implementation be broadly supported, there's the need to stick to the 'old' ARIA 1.0 pattern. 

/Cc @tyxla @mirka

## Why?

The combobox implementation needs to be consistent across all the codebase and use the most broadly supported ARIA pattern.

## How?

- reverts the change from https://github.com/WordPress/gutenberg/pull/43278 changing `aria-controls` back to `aria-owns`
- additionally, improves the usage of `aria-selected`, as it needs to be set only on the highlighted item
- enforces correct usage of these attributes in the tests, to prevent new regressions

## Testing Instructions

- edit a post
- use Safari and VoiceOver
- add a link to a word in a paragraph
- type 'hello' in the URL input search to get some link suggestions
- use the Down and Up keys to navigate through the suggestion options
- observe VoiceOver does announce the options

Optionally test with the most common browser / screen reader combinations. I tested with:
- macOS: Safari + VoiceOver
- Windows: Firefox + NVDA
- Windows: Chrome + NVDA
- Windows: JAWS 2019 + Chrome
- Windows: JAWS 2019 + Firefox

Update: note to clarify that Safari and VoiceOver are buggy anyways. When testing, you may notice some weirdness in the way the options are announced e.g.:

- sometimes they are announced as `{option name}, clickable`
- sometimes they are announced as `{option name}, selected, (1 of n), completion selected` (this is the expected behavior)
- sometimes you may need to delete one of the typed characters and re-type it again to make VoiceOver announce the options

### Testing Instructions for Keyboard

See above.

## Screenshots or screencast 

Screenshot 2023-01-13 at 17 09 33

### CODE REVIEW

**Review by alexstine — APPROVED** [review]
Stupid Voiceover. This is why no one can ever move forward, Apple lags years behind in basic accessibility. It generally is not a problem in native applications but comes back to hurt us big time on the web. So annoying. This has also become much worse in IOS 16 for mobile devices. Wish Apple would bring back their old values.

My political comments out of the way, this looks good from a code perspective. I trust your judgement in testing @afercia, thanks for catching this.

File: packages/block-editor/src/components/link-control/test/index.js [hunk 1]
```diff
@@ -140,6 +140,17 @@ describe( 'Basic rendering', () => {
 		expect( searchInput ).toBeInTheDocument();
 	} );
 
+	it( 'should have aria-owns attribute to follow the ARIA 1.0 pattern', () => {
+		render( <LinkControl /> );
+
+		// Search Input UI.
+		const searchInput = screen.getByRole( 'combobox', { name: 'URL' } );
+
+		expect( searchInput ).toBeInTheDocument();
```
Comment (tyxla) [inline packages/block-editor/src/components/link-control/test/index.js]: Does it make sense to change this assertion to use `toBeVisible()`? I've found this to be a bit more specific query, since `toBeInTheDocument()` will be truthy if the element is not accessible to the user at all. 

```suggestion
		expect( searchInput ).toBeVisible();
```

File: packages/block-editor/src/components/link-control/test/index.js [hunk 2]
```diff
@@ -140,6 +140,17 @@ describe( 'Basic rendering', () => {
...
+		expect( searchInput ).not.toHaveAttribute( 'aria-controls' );
```
Comment (tyxla) [inline packages/block-editor/src/components/link-control/test/index.js:152]: I don't see a good reason to have this assertion in place. I think we can remove it. 

```suggestion
```

**Review by tyxla — APPROVED** [review]
LGTM 👍 Thanks @afercia 🙌 

I've just left a couple of suggestions on the tests, but nothing critical.

File: packages/block-editor/src/components/link-control/test/index.js (hunk 2 above)
Comment (afercia) [reply packages/block-editor/src/components/link-control/test/index.js:152]: I'd like to make sure `aria-selected` is set only on the highlighted item and omitted from all the other items. The current usage is wrong and I'd like to to avoid future misuse. I'd agree it's best to move this to a separate test though.

File: packages/block-editor/src/components/link-control/test/index.js (hunk 1 above)
Comment (afercia) [reply packages/block-editor/src/components/link-control/test/index.js]: Makes sense. There are a few more occurrences of `toBeInTheDocument()`. Some of them use `not` and they actually test for some elements to not be in the DOM. I guess the other ones should be changed to `toBeVisible()`.

File: packages/block-editor/src/components/link-control/test/index.js (hunk 2 above)
Comment (tyxla) [reply packages/block-editor/src/components/link-control/test/index.js:152]: I understand. thanks for clarifying. I guess it's worth adding some context as a comment to the test then 👍

Comment (afercia) [reply packages/block-editor/src/components/link-control/test/index.js:152]: My bad, I see now you mentioned the assertion for `aria-controls`, not the one for `aria-selected`. Will add comments to clarify both.

**Comment by afercia** [comment]
Thanks everyone for the feedback. Everything is green, merging.
